apps/fz_http/lib/fz_http/api_tokens/api_token.ex

```diff
@@ -0,0 +1,51 @@
+defmodule FzHttp.ApiTokens.ApiToken do
+  @moduledoc """
+  Stores API Token metadata to check for revocation.
+  """
+  use FzHttp, :schema
+  import Ecto.Changeset
+
+  @max_per_user 10
+
+  schema "api_tokens" do
+    field :expires_at, :utc_datetime_usec
+
+    # User-friendly way to set expires_at
+    field :expires_in, :integer, virtual: true, default: 30
+
+    belongs_to :user, FzHttp.Users.User
+
+    timestamps(updated_at: false)
+  end
+
+  def changeset(api_token, attrs, opts \\ []) do
+    api_token
+    |> cast(attrs, ~w[
+      user_id
+      expires_in
+      expires_at
+    ]a)
+    |> validate_required([:user_id, :expires_in])
+    |> validate_number(:expires_in, greater_than_or_equal_to: 1, less_than_or_equal_to: 90)
+    |> resolve_expires_at()
+    |> validate_required(:expires_at)
+    |> assoc_constraint(:user)
+    |> maybe_validate_count_per_user(@max_per_user, opts[:count_per_user])
+  end
+
+  def max_per_user, do: @max_per_user
+
+  defp resolve_expires_at(changeset) do
```

Comment:
I feel like this is a UI concept that leaked to the domain changeset function. UI can render any date-picking element for input, but long-term its code should not be part of domain code.

Reply:
I had the same feeling but thought I could sneak this past you ;-)

Reply:
Hmmm, on second thought: things are much cleaner when this is done at the Changeset level. Otherwise we need JS or view logic to convert days to and from the DateTime type. Open to suggestions if you know a better way.

We use a similar pattern for `password` and `password_confirmation` -- i.e. leveraging Ecto.Changeset in the schema to validate virtual attributes.

Reply:
I think you still can handle one-off param before sending it to the changeset, you can even validate it with a schemaless changeset (that will be only single front-end endpoint specific). I know it's more code but at the same time as the app grows larger, it's important to have clear boundaries between contexts and domains.

Reply:
Maybe let's keep it like this for now and decide how to solve it later. It's hard for me to give you an exact example and I don't want you to spend time figuring out the design. Just add XXX here that this is a view concern that we want to move out later.